Project 2

SusanaPerezJ opened the pull request.

SusanaPerezJ committed:
Modify servlet to display my name

SusanaPerezJ committed:
Modify servlet to fetch a message from the server

wait321 commented on portfolio/src/main/webapp/index.html:
```diff
@@ -28,12 +28,16 @@ <h2>Pictures!</h2>
       <p>Click this button to see a random fact about me:</p>
       <button onclick="addRandomFact()">Click Here!</button>
       <div id="fact-container"></div>
+      <br><br/>
```
Did you mean `<br></br>`? In any case br tags just add a line break immediately, so we always do `<br />`, which is the same as `<br></br>`, to save space.

wait321 replied:
Sorry, I should have been more clear. Change to `<br />`. What you changed it to is valid, but it's pointlessly long.

SusanaPerezJ replied:
thank you, I'll make the changes

wait321 reviewed (changes requested):
One minor issue with HTML, but otherwise looks good.

aminry commented on portfolio/src/main/webapp/script.js:
```diff
@@ -26,3 +26,9 @@
+function getFetchedMessage(){
+    fetch('/data').then(response => response.text()).then ((message) => {
+      document.getElementById('message-container').innerText = message;
+    });
```
Also add error handling in case the call fails.

wait321 replied:
Just add
.catch((error) => {
  console.error('Error:', error);
});

SusanaPerezJ replied:
sounds good, thanks!

wait321 replied:
You still need to add the catch

SusanaPerezJ committed:
Add comments feature to the page

SusanaPerezJ committed:
Modify fetch method to fetch comments

Add method to create <li> element containing text

SusanaPerezJ committed:
Modify doGet() method to return comments

Add doPost() method to populate ArrayList

Add getParameter() to return the request parameter

Add validComment() to only add comments that are not empty

SusanaPerezJ committed:
Add comments feature to the page

SusanaPerezJ committed:
Modify fetch method to fetch comments

Add method to create <li> element containing text

SusanaPerezJ committed:
Modify doGet() method to return comments

Add doPost() method to populate ArrayList

Add getParameter() to return the request parameter

Add validComment() to only add comments that are not empty

wait321 commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -19,14 +19,61 @@
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
+import java.util.*;
```
Don't import .* it makes it so difficult to figure out what your dependencies are.

SusanaPerezJ replied:
noted, thank you

wait321 commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -19,14 +19,61 @@
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
+import java.util.*;
+import com.google.gson.Gson;
```
You will need a pom.xml change to download this library before  your code can find it
https://mvnrepository.com/artifact/com.google.code.gson/gson/2.8.6

SusanaPerezJ replied:
I added this dependency to the pom file already. should I commit that change too?

wait321 replied:
Yes. Please add it to the pull request

wait321 commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -19,14 +19,61 @@
+    //Handles empty comments, does not add them to the thread
+    if(validComment(text)){
+        comments.add(text);
+        response.setContentType("text/html");
```
Delete this line. This shouldn't be necessary if you're sending a redirect

wait321 commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -19,14 +19,61 @@
+    String json = new Gson().toJson(comments);
+    response.getWriter().println(json);
+  }
+ /**
```
add a blank line between methods

wait321 commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -19,14 +19,61 @@
+    if(comment.isEmpty()){
+        return false;
+      }
+    for(int j = 0; j < comment.length(); j++){
```
There's a library method that already does this: trim() https://www.w3schools.com/java/showjava.asp?filename=demo_ref_string_trim

wait321 commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -19,14 +19,61 @@
+ * Validates a comment if it has at least a character that is not a white space
+ */
+  public boolean validComment(String comment){
+    if(comment.isEmpty()){
```
this entire method can be replaced with the line

return !comment.trim().isEmpty();

wait321 commented on portfolio/src/main/webapp/index.html:
```diff
@@ -28,12 +28,22 @@ <h2>Pictures!</h2>
+      <p>Leave a comment below!</p>
+        <textarea name="comment-box" rows="6" cols="100"></textarea>
+        <p><class="space-between-elements"></p>
+        <input type="submit">
```
Need to close the tag. Also, this looks like a button, which is easier to understand if you use `<button type="submit" />`

wait321 commented on portfolio/src/main/webapp/index.html:
```diff
@@ -28,12 +28,22 @@ <h2>Pictures!</h2>
+      <form action="/data" method="POST" class="space-between-elements">
+      <p>Leave a comment below!</p>
+        <textarea name="comment-box" rows="6" cols="100"></textarea>
+        <p><class="space-between-elements"></p>
```
There's no HTML tag `<class ...>`. You can just add the class to textarea or the p above. Example:
`<textarea class="space-between-elements" ...`

wait321 commented on portfolio/src/main/webapp/script.js:
```diff
@@ -14,15 +14,34 @@
      const randomFact = facts[Math.floor(Math.random() * facts.length)];
      
-     //Add random fact to the page
+     //Add it to the page
```
revert this change. "random fact" is easier to understand than "it".

SusanaPerezJ replied:
the reason i changed it was because on the comment of the method i already mentioned that is adding a random fact so i didn't want to sound repetitive. 
Do you think it's better to revert that change or just delete the method comment altogether?

wait321 replied:
Either is fine. I don't feel any strong preference between the 2 options.

aminry commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -19,14 +19,61 @@
-    response.getWriter().println("<h1>Hello world!</h1>");
+    response.setContentType("application/json");
+    String json = new Gson().toJson(comments);
+    response.getWriter().println(json);
```
Have member variable for this gson instance, and don't create a new instance everytime.

aminry replied:
nit: no need to use println(, instead simply use print(

aminry commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -19,14 +19,61 @@
+ */
+  private String getParameter(HttpServletRequest request, String name, String defaultValue) {
+    String value = request.getParameter(name);
+    if (value == null) {
```
nit: use ternary operator

return value == null ? defaultValue : value;

SusanaPerezJ committed:
Add Maven dependency

SusanaPerezJ committed:
Add error handling to fetch function in getCommentThread()

SusanaPerezJ committed:
Modify getPArameter() method to use ternary operator

Modify validComment() method

Modify Gson instance to be a member variable

SusanaPerezJ committed:
Modify submit button

SusanaPerezJ committed:
Add space-between-elements class

aminry commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -19,14 +19,49 @@
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
```
nit: Please sort the import alphabetical

aminry commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -19,14 +19,49 @@
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
+  public ArrayList<String> comments = new ArrayList<>();
+  private Gson gson = new Gson();
```
Please make this "static final"

aminry commented on portfolio/src/main/java/com/google/sps/servlets/DataServlet.java:
```diff
@@ -19,14 +19,49 @@
+/** Servlet that returns thread of comments. */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
+  public ArrayList<String> comments = new ArrayList<>();
```
Please make this static final, and it can be private.

aminry commented on portfolio/src/main/webapp/index.html:
```diff
@@ -28,12 +28,21 @@ <h2>Pictures!</h2>
       <a href="https://www.linkedin.com/in/susana-perez-juarez/">LinkedIn profile</a>
       <a href="https://github.com/SusanaPerezJ">GitHub profile</a>
+      <form action="/data" method="POST" class="space-between-elements">
+      <p>Leave a comment below!</p>
```
nit: you want tap this forward to be consistent

wait321 replied:
All lines between an open tag like `<form>` and its closing tag `</form>` should have extra spaces in front (like what you did for the line after `<body>`)

aminry commented on portfolio/src/main/webapp/index.html:
```diff
@@ -28,12 +28,21 @@ <h2>Pictures!</h2>
+      <p>Leave a comment below!</p>
+        <textarea name="comment-box" rows="6" cols="100"></textarea>
+        <button type="submit">Submit comment</button>
+        <p>Comments:</p>  
```
This portion does not need to be inside the form tag.

SusanaPerezJ committed:
Modify elements inside form

SusanaPerezJ committed:
Modify doPost() method

Modify member variables to be static final
